### PR TITLE
Update `SSL_SERIALIZED_SESSION_CONFIG_BITFLAG` with new flags

### LINKED ISSUE #8686: Outdated SSL_SERIALIZED_SESSION_CONFIG_BITFLAG in ssl_tls.c
### Summary
The TLS session serialized data (saved with mbedtls_ssl_session_save() and loaded with mbedtls_ssl_session_load()) depend on the configuration thus session serialized data saved with one configuration of the library may not be used by the library compiled under another configuration. To detect that, before to parse serialized data, mbedtls_ssl_session_load() check that the mask of configurations 
saved in the data in equal to the mask of configurations of the library, namely SSL_SERIALIZED_SESSION_CONFIG_BITFLAG, trying to load the data. SSL_SERIALIZED_SESSION_CONFIG_BITFLAG has not been updated when adding data for early data and record size limit. Other compilation flags may be missing as well.

### System information
Mbed TLS version (number or commit id): development, post 3.5 release.
Operating system and version: -
Configuration (if not default, please attach `mbedtls_config.h`): -
Compiler and options (if you used a pre-built binary, please indicate how you obtained it): -
Additional environment information: -

### Expected behavior
SSL_SERIALIZED_SESSION_CONFIG_BITFLAG in session serialized data reflects correctly the enabled configuration options for the library that saved the session serialized data.

### Steps to reproduce
This has been undetected because we do not have test covering that. We should add tests in test_suite_ssl where we save a session and then check SSL_SERIALIZED_SESSION_CONFIG_BITFLAG. With several test cases with the same test function but different dependencies (based on the configuration options relevant to session serialized data) we should be able to add tests that would potentially detect when we miss updating SSL_SERIALIZED_SESSION_CONFIG_BITFLAG while changing session serialized data.

### PR DESCRIPTION
**Pull request by davidhorstmann-arm**
Fixes #8686

Add missing flags to `SSL_SERIALIZED_SESSION_CONFIG_BITFLAG`. Previously, the bitflag did not uniquely identify a possible structure of a serialized session, meaning that multiple different incompatible session structures could be serialized with the same header.

The options added are as follows:
* From TLS 1.3 work:
    * Server name indication
    * Early data
    * Record size limit
* Preexisting in TLS 1.2 code:
    * Keep peer certificate

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided
- [x] **backport** #8823 
- [x] **tests** not required - tests will be provided in a followup PR

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

Help make review efficient:
* Multiple simple commits
  - please structure your PR into a series of small commits, each of which does one thing
* Avoid force-push
  - please do not force-push to update your PR - just add new commit(s)
* See our [Guidelines for Contributors](https://mbed-tls.readthedocs.io/en/latest/reviews/review-for-contributors/) for more details about the review process.

### CODE REVIEW

**Review by mpg — COMMENTED** [review]
The changes here are looking good, so I'll mostly comment on pre-existing stuff :)

1. The way functions are placed in the file makes no sense to me: `ssl_session_save()` on line 3751, `ssl_tls12_session_save()` on line 8973 (so it has to be forward declared), `ssl_tls13_session_save()` on line 2476. (Line numbers before this PR, but it's not about the exact numbers, just that they're in vastly different places.)
2. About comments, I can understand why they're this way, but I think it would be better to have all the format description in one place - that is, both the 1.2 and 1.3 structure documented above `ssl_session_save()` together with the header in order to get everything in one place - then add comments to the version-specific save() and load() functions referring to that.
3. Then in the format description I think it would be nice to have `the #if`s so that we can have a clear view of what influenced the format how.

I feel quite strongly about (1) but less so about (2) and (3) though I think we should try to get things more organized even if it's not in the way I'm suggesting.

**Comment by davidhorstmann-arm** [comment]
I believe I've addressed all of the feedback here, should be ready for re-review.

**Comment by davidhorstmann-arm** [comment]
However, since what was reviewed was fairly small, I will rebase and resolve conflicts now.

**Comment by davidhorstmann-arm** [comment]
Rebase done, this is ready for re-review

**Comment by davidhorstmann-arm** [comment]
@mpg Would you like me to duplicate the refactoring changes in the backport as well?

**Comment by mpg** [comment]
> @mpg Would you like me to duplicate the refactoring changes in the backport as well?

Yes please :)

**Review by mpg — DISMISSED** [review]
LGTM, thanks!

**Comment by tom-cosgrove-arm** [comment]
I'm seeing something a bit weird: GitHub says `+4,950 −4,880` at the top, and if I check out the PR, I see this.

Any idea what's going on?

```
$ git branch --show-current
fix-config-bitflag

$ git log -n1
commit 531aca28103063dc75066346ee2a1e2bc2272218 (HEAD -> fix-config-bitflag)
Author: David Horstmann <david.horstmann@arm.com>
Date:   Thu Feb 29 18:14:28 2024 +0000

    Fix missing fields in ssl session struct comment

    The endpoint and version were factorized out into the main session.
    Update the session struct comment to reflect these new fields, as was
    previously missed.

    Signed-off-by: David Horstmann <david.horstmann@arm.com>

$ git log --oneline -n7
531aca281 (HEAD -> fix-config-bitflag) Fix missing fields in ssl session struct comment
cb01b361e Move session descriptions into a single comment
80a966876 Add config guards to session struct comments
e59f970f2 Move session functions to same part of file
92b258bb5 Update ssl session serialization config bitflag
5c5a32f52 Add session config bit for KEEP_PEER_CERTIFICATE
9b4e964c2 Merge pull request #8760 from ronald-cron-arm/tls13-write-early-data

$ git show HEAD~3 | wc
   10993   40422  406931

$ git log -n1 HEAD~3
commit e59f970f2874260011f328085c2fbcdb31fb02fb
Author: David Horstmann <david.horstmann@arm.com>
Date:   Thu Feb 29 16:08:57 2024 +0000

    Move session functions to same part of file

    Ensure that session save and load functions are not scattered
    throughout ssl_tls.c but are in the same part of the file.

    Signed-off-by: David Horstmann <david.horstmann@arm.com>
```

**Comment by mpg** [comment]
Not sure what's happening on github or your machine:
```
% git diff --stat development..531aca281 | tail -n1
1 file changed, 731 insertions(+), 661 deletions(-)
% % git show e59f970f2 | wc -l
1360
```

**Comment by tom-cosgrove-arm** [comment]
Okay, the large commit (`e59f970f2874260011f328085c2fbcdb31fb02fb`) is "Move session functions to same part of file". The difference in diff stats depends on the defaults set for git diff. I have recreated the change in that commit with a script, and am satisfied that it's deleting (original) lines 434..437 and moving serialisation of TLS 1.2 and 1.3 sessions together, in that order.

File: library/ssl_tls.c [hunk 1]
```diff
@@ -3640,5693 +3360,6043 @@ int mbedtls_ssl_get_session(const mbedtls_ssl_context *ssl,
...
+#define SSL_SERIALIZED_SESSION_CONFIG_CRT 0
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
 
-#if defined(MBEDTLS_X509_CRT_PARSE_C)
-    ssl_clear_peer_cert(session);
-#endif
+#if defined(MBEDTLS_SSL_KEEP_PEER_CERTIFICATE)
+#define SSL_SERIALIZED_SESSION_KEEP_PEER_CRT 1
+#else
+#define SSL_SERIALIZED_SESSION_KEEP_PEER_CRT 0
```
Comment (tom-cosgrove-arm) [inline library/ssl_tls.c:4012]: (minor, naming) All the other flags have `CONFIG` in their name, which would make this `SSL_SERIALIZED_SESSION_CONFIG_KEEP_PEER_CRT`

File: library/ssl_tls.c [hunk 2]
```diff
@@ -3640,5693 +3360,6043 @@ int mbedtls_ssl_get_session(const mbedtls_ssl_context *ssl,
...
-#define SSL_SERIALIZED_CONTEXT_CONFIG_DTLS_BADMAC_LIMIT_BIT     1
-#define SSL_SERIALIZED_CONTEXT_CONFIG_DTLS_ANTI_REPLAY_BIT      2
-#define SSL_SERIALIZED_CONTEXT_CONFIG_ALPN_BIT                  3
+#define SSL_SERIALIZED_SESSION_CONFIG_TIME_BIT          0
+#define SSL_SERIALIZED_SESSION_CONFIG_CRT_BIT           1
+#define SSL_SERIALIZED_SESSION_CONFIG_CLIENT_TICKET_BIT 2
+#define SSL_SERIALIZED_SESSION_CONFIG_MFL_BIT           3
+#define SSL_SERIALIZED_SESSION_CONFIG_ETM_BIT           4
+#define SSL_SERIALIZED_SESSION_CONFIG_TICKET_BIT        5
+#define SSL_SERIALIZED_SESSION_CONFIG_KEEP_CRT_BIT      6
```
Comment (tom-cosgrove-arm) [inline library/ssl_tls.c:4063]: (minor, naming) This one would more properly be `SSL_SERIALIZED_SESSION_CONFIG_KEEP_PEER_CRT_BIT`

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM with a couple of naming inconsistencies (up to you if you want to fix them)

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM, thanks

**Comment by davidhorstmann-arm** [comment]
Sorry, missed something small. Now ready for review

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Review by tom-cosgrove-arm — APPROVED** [review]

**Review by ronald-cron-arm — APPROVED** [review]

**Comment by gilles-peskine-arm** [comment]
@tom-cosgrove-arm `git show --color-moved --histogram e59f970f2874260011f328085c2fbcdb31fb02fb` looks a lot neater than `git show --color-moved e59f970f2874260011f328085c2fbcdb31fb02fb` (at least with my version of Git). In general I've found that `--histogram` tends to give better results than the default diff algorithm, at least for the kind of patches we tend to have.

**Comment by tom-cosgrove-arm** [comment]
CI is green, so removing `needs-ci` label

**Comment by tom-cosgrove-arm** [comment]
This is approved, but backport is still being reviewed

**Pull request merged**